encodings/internal.js

```diff
@@ -196,3 +198,34 @@ InternalDecoderCesu8.prototype.end = function() {
         res += this.defaultCharUnicode;
     return res;
 }
+
+//------------------------------------------------------------------------------
+// check the chunk boundaries for surrogate pair
+
+function InternalEncoderUtf8(options, codec) {
+    this.lowSurrogate = '';
+}
+
+InternalEncoderUtf8.prototype.write = function (str) {
+    var lastr = str[str.length - 1] || '',
+        charCode = lastr.charCodeAt(0)
+    if (55296 < charCode && charCode <= 56319) {
+        if (this.lowSurrogate) {
+            str = this.lowSurrogate + str
+        }
+        this.lowSurrogate = lastr
+        str = str.slice(0, str.length - 1)
+    } else {
+        str = this.lowSurrogate + str
+        this.lowSurrogate = ''
+    }
+    return Buffer.from(str, this.enc);
```

Comment:
I see you're returning Buffer.from in the end, this is good.
The `lastr` handling became more messy though. I thought we could do something like this:
```
InternalEncoderUtf8.prototype.write = function (str) {
    if (!str) return;

    if (this.lowSurrogate) {
        str = this.lowSurrogate + str;
        this.lowSurrogate = '';
    }

    var lastCharCode = str.charCodeAt(str.length-1);
    if (0xD800 < lastCharCode && lastCharCode <= 0x...) {
        this.lowSurrogate = str.slice(str.length - 1);
        str = str.slice(0, str.length - 1)
    }

    return Buffer.from(str, this.enc);
}
```